src/utils.ts

```diff
@@ -16,32 +33,137 @@ export const generateTSSEndpoints = (tssNodeEndpoints: string[], parties: number
+ * @param shareMnemonic - The mnemonic to convert.
+ * @returns The factor key.
+ */
+export async function mnemonicToKey(tKey: ThresholdKey, shareMnemonic: string): Promise<BN> {
+  const factorKey = await (tKey.modules.shareSerialization as ShareSerializationModule).deserialize(shareMnemonic, "mnemonic");
+  return factorKey;
+}
+
+/**
+ * Converts an arbitrary string to a factor key over an elliptic curve.
+ * @param ec - The elliptic curve.
+ * @param shareMnemonic - The mnemonic to convert.
+ * @returns The factor key.
+ */
+export async function stringToKey(ec: EC.curve.base, s: string): Promise<BN> {
+  const buf = Buffer.from(s);
+  // TODO this is compatible with the implementation before, but instead would
+  // be better to hash to scalar instead?
+  const bn = new BN(buf);
+  return bn.mod(ec.n);
+}
+
+/**
+ * Refreshes TSS shares. Allows to change number of shares. New user shares are
+ * only produced for the target indices.
+ * @param tKey - Tkey instance to use.
+ * @param factorPubs - Factor pub keys after refresh.
+ * @param tssIndices - Target tss indices to generate new shares for.
+ * @param factorKeyForExistingTSSShare - Factor key for existing TSS share.
+ * @param signatures - Signatures for authentication against RSS servers.
+ */
+async function refreshTssShares(
+  tKey: ThresholdKey,
+  factorPubs: Point[],
+  tssIndices: number[],
+  factorKeyForExistingTSSShare: BN,
+  signatures: string[]
+) {
+  const { tssShare, tssIndex } = await tKey.getTSSShare(factorKeyForExistingTSSShare);
+
+  const rssNodeDetails = await tKey._getRssNodeDetails();
+  const { serverEndpoints, serverPubKeys, serverThreshold } = rssNodeDetails;
+  const randomSelectedServers = randomSelection(
+    new Array(rssNodeDetails.serverEndpoints.length).fill(null).map((_, i) => i + 1),
+    Math.ceil(rssNodeDetails.serverEndpoints.length / 2)
+  );
+
+  const verifierNameVerifierId = tKey.serviceProvider.getVerifierNameVerifierId();
+  await tKey._refreshTSSShares(true, tssShare, tssIndex, factorPubs, tssIndices, verifierNameVerifierId, {
+    selectedServers: randomSelectedServers,
+    serverEndpoints,
+    serverPubKeys,
+    serverThreshold,
+    authSignatures: signatures,
+  });
+}
+
+export async function addFactorAndRefresh(
+  tKey: ThresholdKey,
+  newFactorPub: Point,
+  newFactorTSSIndex: number,
+  factorKeyForExistingTSSShare: BN,
+  signatures: string[]
+) {
+  if (!tKey) {
+    throw new Error("tkey does not exist, cannot add factor pub");
+  }
+  if (VALID_TSS_INDICES.indexOf(newFactorTSSIndex) === -1) {
+    throw new Error(`invalid new share index: must be one of ${VALID_TSS_INDICES}`);
+  }
+  if (!tKey.metadata.factorPubs || !Array.isArray(tKey.metadata.factorPubs[tKey.tssTag])) {
+    throw new Error(`factorPubs for tssTag = "${tKey.tssTag}" does not exist`);
+  }
+
+  const existingFactorPubs = tKey.metadata.factorPubs[tKey.tssTag];
+  const updatedFactorPubs = existingFactorPubs.concat([newFactorPub]);
```

Comment:
i remember similar functions are there in tkey, why we need custom implementations here?

Reply:
@ieow